10-angular/ui_router/readme.md

```diff
@@ -0,0 +1,581 @@
+
+With all our ducks in a row, we're now ready to make this module actually do something.
+
+# Making this module actually do something
+
+TODO: $locationProvider for clean URLs
+https://scotch.io/quick-tips/pretty-urls-in-angularjs-removing-the-hashtag
+
+We're going to configure this app to have routes to multiple views: a view for the Grumbles index, and a Grumbles show page.
+
+Remember in Express we had a `config/routes.js` file with all of the routes defined in it. Here, we put the `routes` inside a `config` module.
+
+Add `.config` to your `app.js`:
+
+```js
+"use strict";
+
+(function(){
+  angular
+  .module("grumblr", [
+    "ui.router"
+  ])
+  .config([
+    "$stateProvider"
+  ]);
+}());
+```
+
+You'll notice `$stateProvider` is in quotes. You'll also notice that if you refresh the page, you'll get an error. That's OK -- we're going to talk about `$stateProvider` first and then we'll address the error.
+
+## $stateProvider
+
+The `ui.router` module we're requiring includes factories, services, providers, and more. We're not going to talk about the differences between those for now. Just know that `$stateProvider` is something `ui.router` gives us to manage **states**.
+
+A state in Angular is basically a route: it's an umbrella term for a URL, the view associated with it, and any controllers used in that view.
+
+## Config needs functions
+
+The reason `.config` is throwing an error is because Angular expects calling `.config` to always result in it calling a function -- and we haven't yet told `.config` to call a function. The purpose of the function is to actually *do* the configuring.
+
+Here's how we're going to add the function:
+
+```js
+"use strict";
+
+(function(){
+  angular
+  .module("grumblr", [
+    "ui.router"
+  ])
+  .config([
+    "$stateProvider",
+    RouterFunction
+  ]);
+
+  function RouterFunction($stateProvider){
+    
+  }
+}());
+```
+
+## Wat
+
+Two weird things here:
+
+**Weird thing number 1:** First `$stateProvider` is in quotes. Then it isn't. This is how Angular wrangles dependency injection. The reason it does things this goofy way is to protect against minification: compressing your Javascript code to make the file smaller. When you minify Javascript, variable names get changed, but strings don't. Without this, `$stateProvider` would get replaced with a variable name like `c`, which would throw an error because Angular doesn't know about anything named `c`.
+
+Seem like a hack? That's because it is -- a systemic hack.
+
+> "Under the hood, most critical software you use every day (like Mac OS X, or Facebook) contains a terrifying number of hacks and shortcuts that happen to barely fit together into a working whole. It would be like taking apart a brand-new 747 and discovering that the fuel line is held in place by a coat-hanger and the landing gear is attached with duct tape." - Ben Cherry
```

Comment:
I don't feel comfortable with this quote.  More than that, I take offense with it.  My teammates and I work very hard to ensure our code is not hacked together with duct tape.  We are quite proud of writing eloquent code.  There are usually some areas where we can say "there be dragons", but we work to ensure they are encapsulated and obvious.